cookbook/case_studies/feature_engineering/feast_integration/feast_wf.py

```diff
@@ -0,0 +1,351 @@
+#    * - ``FeatureView``
+#      - A FeatureView defines a logical grouping of serveable features.
+#    * - ``FileSource``
+#      - File data sources allow for the retrieval of historical feature values from files on disk for building training datasets, as well as for materializing features into an online store.
+#    * - ``fs.apply()``
+#      - Register objects to metadata store and update related infrastructure.
+#    * - ``get_historical_features()``
+#      - Enrich an entity dataframe with historical feature values for either training or batch scoring.
+@task
+def store_offline(parquet_file: FlyteFile, repo_path: str) -> (str, str):
+    fs = FeatureStore(repo_path=repo_path)
+    horse_colic_entity = Entity(name="Hospital Number", value_type=ValueType.STRING)
+
+    horse_colic_feature_view = FeatureView(
+        name="horse_colic_stats",
+        entities=["Hospital Number"],
+        features=[
+            Feature(name="rectal temperature", dtype=ValueType.FLOAT),
+            Feature(name="total protein", dtype=ValueType.FLOAT),
+            Feature(name="peripheral pulse", dtype=ValueType.FLOAT),
+            Feature(name="surgical lesion", dtype=ValueType.STRING),
+            Feature(name="abdominal distension", dtype=ValueType.FLOAT),
+            Feature(name="nasogastric tube", dtype=ValueType.STRING),
+            Feature(name="outcome", dtype=ValueType.STRING),
+            Feature(name="packed cell volume", dtype=ValueType.FLOAT),
+            Feature(name="nasogastric reflux PH", dtype=ValueType.FLOAT),
+        ],
+        input=FileSource(
+            path=str(parquet_file),
+            event_timestamp_column="timestamp",
+        ),
+        ttl=timedelta(days=1),
+    )
+
+    fs.apply([horse_colic_entity, horse_colic_feature_view])
+    return repo_path, horse_colic_feature_view.name
+
+
+@task
+def retrieve_offline(repo_path: str) -> pd.DataFrame:
+    fs = FeatureStore(repo_path=repo_path)
+    entity_df = pd.DataFrame.from_dict(
+        {
+            "Hospital Number": [
+                "530101",
+                "5290409",
+                "5291329",
+                "530051",
+                "529518",
+                "530101",
+                "529340",
+                "5290409",
+                "530034",
+            ],
+            "event_timestamp": [
+                datetime(2021, 6, 25, 16, 36, 27),
+                datetime(2021, 6, 25, 16, 36, 27),
+                datetime(2021, 6, 25, 16, 36, 27),
+                datetime(2021, 6, 25, 16, 36, 27),
+                datetime(2021, 6, 25, 16, 36, 27),
+                datetime(2021, 7, 5, 11, 36, 1),
+                datetime(2021, 6, 25, 16, 36, 27),
+                datetime(2021, 7, 5, 11, 50, 40),
+                datetime(2021, 6, 25, 16, 36, 27),
+            ],
+        }
+    )
+
+    retrieval_job = fs.get_historical_features(
```

Comment:
where are these historical features stored?

Reply:
For now, locally.
```
project: feature_engineering
registry: data/registry.db
provider: local
online_store:
    path: data/online_store.db
```